cc3d/player5/Plugins/ViewManagerPlugins/DemoBrowser.py

```diff
@@ -390,22 +231,273 @@ def selectDemo(self, selected):
+
+    # Start at location of python.exe and check in every
+    # parent directory for Demos until we hit the root.
+    startDir = Path(__file__)
+
+    for currPath in [startDir] + list(startDir.parents):
+        demosPath = currPath.joinpath("CompuCell3D/Demos")
+        if demosPath.exists():
+            Configuration.setSetting("DemosPath", str(demosPath))
+            return demosPath
+
+    return None  # No demos here!
+
+
+def getDemoList(demo_root_path:Optional[Path]=None) -> list[Path]:
+    glob_results = demo_root_path.rglob("*.cc3d")
+    glob_results = [p.relative_to(demo_root_path) for p in glob_results]
+
+    return glob_results
+
+
+def formatDemoName(name) -> str:
+    outStr = name[0].upper()
+    prev = "A"
+    i = 1
+    while i < len(name):
+        c = name[i]
+        if (c == "_" or c == "-") and prev != " ":
+            c = " "
+        elif c.islower() and prev == " ":
+            c = c.upper()
+        elif c.isupper():
+            if prev.islower() or (i + 1 < len(name) and name[i + 1].islower()):
+                outStr += " "
+
+        outStr += c
+        prev = c
+        i += 1
+    return outStr
+
+
+def splitByUppercase(word):
+    # Extract individual words from camel/pascal case tokens.
+    # Example: "lambdaVolume" splits into "lambda" and "Volume".
+    start = 0
+    i = 1
+    while i <= len(word):
+        if word[i - 1].islower() and (i == len(word) or word[i].isupper()):
+            if i - start < len(word):
+                yield word[start:i]
+            start = i
+        i += 1
+
+
+def tokenizeAllFiles(demoAbsPath):
+    methods = (("*.py", tokenizePython), ("*.xml", tokenizeXml), (DESCRIPTION_FNAME, tokenizeText))
+    for fileType, tokenizer in methods:
+        for filePath in demoAbsPath.parent.rglob(fileType):
+            with open(filePath, "r") as fp:
+                for line in fp.readlines():
+                    for word in tokenizer(line):
+                        if word:
+                            yield word.lower()
+                            for subword in splitByUppercase(word):
+                                yield subword.lower()
+
+
+def tokenizePython(line):
+    """
+    Generates each variable or method name from a line of Python code.
+    It is not perfect.
+    """
+    for word in line.split():
+        if "." in word:
+            word = re.sub("[^a-zA-Z \n\.]", ".", word)  # Remove numbers and special chars
+            for varName in word.split("."):
+                yield varName
+
+
+def tokenizeXml(line):
+    """
+    Generates all the key words (XML tag names and parameter names)
+    from a line of XML.
+    """
+    if "!--" in line:  # exclude comments, even if they have useful tags
+        return
+    for word in re.split("[^a-zA-Z]", line):
+        yield word
+
+
+def tokenizeText(line):
+    for word in re.split("[^a-zA-Z]", line):
+        yield word
+
+
+def preIndexSearchResults(demo_root_path:Optional[Path]=None):
```

Comment:
I am fine with this change, but I feel that it adds complexity to the code.

Reply:
It actually makes multithreading possible Otherwise you will be getting this type of error:
```
sqlite3.ProgrammingError: SQLite objects created in a thread can only be used in that same thread. The object was created in thread id 44524 and this is thread id 29580.
```